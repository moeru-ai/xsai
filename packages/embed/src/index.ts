export { embed as default, embed, type EmbedOptions, type EmbedResponseUsage, type EmbedResult } from './utils/embed'
export { embedMany, type EmbedManyOptions, type EmbedManyResult } from './utils/embed-many'
