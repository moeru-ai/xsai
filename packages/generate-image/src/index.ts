import type { CommonRequestOptions } from '@xsai/shared'

import { requestBody, requestHeaders, requestURL, responseCatch, responseJSON } from '@xsai/shared'

export interface GenerateImageOptions extends CommonRequestOptions {
  [key: string]: unknown
  /**
   * The number of images to generate.
   * @default `1`
   */
  n?: number
  /** A text description of the desired image(s). */
  prompt: string
  /**
   * The format in which generated images with `dall-e-2` and `dall-e-3` are returned from the API.
   * `gpt-image-1` only supports `b64_json`, and `url` will be ignored.
   *
   * Images will always be Base64-encoded in the result. If `url` is used, the images will be
   * fetched upon receiving the response.
   */
  responseFormat?: 'b64_json' | 'url'
  /**
   * The size of the generated images.
   * @default `1024x1024`
   */
  size?: `${number}x${number}`
}

export interface GenerateImageResponse {
  created: number
  data: {
    /**
     * The Bse64-encoded JSON of the generated image. Default value for `gpt-image-1`, and only
     * present if `response_format` is set to `b64_json` for `dall-e-2` and `dall-e-3`.
     */
    b64_json?: string
    /**
     * For `dall-e-3` only, the revised prompt that was used to generate the image.
     */
    revised_prompt?: string
    /**
     * When using `dall-e-2` or `dall-e-3`, the URL of the generated image if `response_format` is
     * set to `url` (default value). Unsupported for `gpt-image-1`.
     */
    url?: string
  }[]
}

export interface GenerateImageResult {
  image: GenerateImageResultImage
  images: GenerateImageResultImage[]
}

export interface GenerateImageResultImage {
  base64: string
  mimeType: string
}

/** @internal */
const mimeTypes = {
  '/9j/': 'image/jpg',
  'AAAAIGZ0eXBhdmlm': 'image/avif',
  'iVBORw0KGgo': 'image/png',
  'R0lGOD': 'image/gif',
  'UklGRg==': 'image/webp',
}

/** @internal */
const convertImage = (b64_json: string) => {
  const key = Object.keys(mimeTypes).find(prefix => b64_json.startsWith(prefix)) as keyof typeof mimeTypes | undefined
  const mimeType = mimeTypes[key ?? 'iVBORw0KGgo']

  return {
    // eslint-disable-next-line @masknet/string-no-data-url
    base64: `data:${mimeType};base64,${b64_json}`,
    mimeType,
  }
}

/** @internal */
const responseBlobAsDataURL = async (res: Response): Promise<string> =>
  responseCatch(res)
    .then(async (res) => {
      const blob = await res.blob()

      try {
        return await new Promise<string>((resolve, reject) => {
          const reader = new FileReader()
          reader.onloadend = () => resolve(reader.result as string)
          reader.onerror = reject
          reader.readAsDataURL(blob)
        })
      }
      catch {
        throw new Error(`Failed to parse response blob, response URL: ${res.url}`)
      }
    })

/** @experimental */
export const generateImage = async (options: GenerateImageOptions): Promise<GenerateImageResult> =>
  (options.fetch ?? globalThis.fetch)(requestURL('images/generations', options.baseURL), {
    body: requestBody(options),
    headers: requestHeaders({
      'Content-Type': 'application/json',
      ...options.headers,
    }, options.apiKey),
    method: 'POST',
    signal: options.abortSignal,
  })
    .then(responseJSON<GenerateImageResponse>)
    .then(async ({ data }) =>
      Promise.all(
        data.map(async (img, i) => {
          if (typeof img.b64_json === 'string') {
            return convertImage(img.b64_json)
          }

          if (typeof img.url === 'string') {
            return (options.fetch ?? globalThis.fetch)(new URL(img.url), {
              signal: options.abortSignal,
            })
              .then(responseBlobAsDataURL)
              .then((dataURL) => {
                const sepIndex = dataURL.indexOf(';')
                const mimeType = dataURL.substring(5, sepIndex) // `data:(...);...`
                return {
                  base64: dataURL,
                  mimeType,
                }
              })
          }

          throw new Error(`Unrecognized image at index ${i}: ${JSON.stringify(img)}`)
        }),
      ),
    )
    .then(images => ({ image: images[0], images }))
