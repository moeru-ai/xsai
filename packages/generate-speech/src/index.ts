import { type CommonRequestOptions, requestBody, requestHeaders, requestURL, responseCatch } from '@xsai/shared'

export interface GenerateSpeechOptions extends CommonRequestOptions {
  [key: string]: unknown
  input: string
  /** @default `mp3` */
  responseFormat?: 'aac' | 'flac' | 'mp3' | 'opus' | 'pcm' | 'wav'
  /** @default `1.0` */
  speed?: number
  voice: string
}

/**
 * @experimental WIP
 * @param options generate speech options.
 * @returns audio array buffer
 */
export const generateSpeech = async (options: GenerateSpeechOptions): Promise<ArrayBuffer> =>
  await (options.fetch ?? globalThis.fetch)(requestURL('audio/speech', options.baseURL), {
    body: requestBody(options),
    headers: requestHeaders({
      'Content-Type': 'application/json',
      ...options.headers,
    }, options.apiKey),
    method: 'POST',
    signal: options.abortSignal,
  })
    .then(responseCatch)
    .then(res => res.arrayBuffer())

export default generateSpeech
