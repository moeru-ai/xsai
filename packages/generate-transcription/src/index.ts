import type { CommonRequestOptions } from '@xsai/shared'

import { requestHeaders, requestURL, responseJSON } from '@xsai/shared'

export interface GenerateTranscriptionOptions extends CommonRequestOptions {
  file: Blob
  fileName?: string
  language?: string
  prompt?: string
  temperature?: string

  // response_format?: 'json' | 'text' | 'srt' | 'verbose_json' | 'vtt'
  // timestamp_granularities?: ('segment' | 'word')[]
}

export interface GenerateTranscriptionResult {
  text: string
}

export const generateTranscription = async (options: GenerateTranscriptionOptions): Promise<GenerateTranscriptionResult> => {
  const body = new FormData()

  body.append('model', options.model)
  body.append('file', options.file, options.fileName)
  options.language != null && body.append('language', options.language)
  options.prompt != null && body.append('prompt', options.prompt)
  options.temperature != null && body.append('temperature', options.temperature)

  return (options.fetch ?? globalThis.fetch)(requestURL('audio/transcriptions', options.baseURL), {
    body,
    headers: requestHeaders(options.headers, options.apiKey),
    method: 'POST',
    signal: options.abortSignal,
  })
    .then(responseJSON<GenerateTranscriptionResult>)
    .then(({ text }) => ({ text: text.trim() }))
}
