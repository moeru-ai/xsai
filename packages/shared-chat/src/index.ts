export * from './error'
export type * from './types'
export * from './utils'
