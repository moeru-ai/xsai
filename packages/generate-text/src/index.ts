import type { AssistantMessageResponse, ChatOptions, FinishReason, Message, Tool, Usage } from '@xsai/shared-chat'

import { chat } from '@xsai/shared-chat'

export interface GenerateTextOptions extends ChatOptions {
  /** @default 1 */
  maxSteps?: number
  onStepFinish?: (step: GenerateTextStepResult) => Promise<void> | void
  /** @internal */
  steps?: GenerateTextStepResult[]
  /** if you want to enable stream, use `@xsai/stream-{text,object}` */
  stream?: never
  tools?: Tool[]
}

export interface GenerateTextResponse {
  choices: {
    finish_reason: FinishReason
    index: number
    message: AssistantMessageResponse
  }[]
  created: number
  id: string
  model: string
  object: 'chat.completion'
  system_fingerprint: string
  usage: Usage
}

export interface GenerateTextResult {
  finishReason: FinishReason
  messages: Message[]
  steps: GenerateTextStepResult[]
  text?: string
  toolCalls: GenerateTextToolCall[]
  toolResults: GenerateTextToolResult[]
  usage: Usage
}

export interface GenerateTextStepResult {
  finishReason: FinishReason
  stepType: 'continue' | 'initial' | 'tool-result'
  text?: string
  toolCalls: GenerateTextToolCall[]
  toolResults: GenerateTextToolResult[]
  usage: Usage
}

export interface GenerateTextToolCall {
  args: string
  toolCallId: string
  toolCallType: 'function'
  toolName: string
}

export interface GenerateTextToolResult {
  args: Record<string, unknown>
  result: string
  toolCallId: string
  toolName: string
}

/** @internal */
type RawGenerateText = (options: GenerateTextOptions) => RawGenerateTextTrampoline<GenerateTextResult>

/** @internal */
type RawGenerateTextTrampoline<T> = Promise<(() => RawGenerateTextTrampoline<T>) | T>

/** @internal */
const rawGenerateText: RawGenerateText = async (options: GenerateTextOptions) =>
  chat({
    ...options,
    maxSteps: undefined,
    messages: options.messages,
    steps: undefined,
    stream: false,
  })
    .then(async res => res.json() as Promise<GenerateTextResponse>)
    .then(async ({ choices, usage }) => {
      const messages: Message[] = structuredClone(options.messages)
      const steps: GenerateTextStepResult[] = options.steps ? structuredClone(options.steps) : []
      const toolCalls: GenerateTextToolCall[] = []
      const toolResults: GenerateTextToolResult[] = []

      const { finish_reason: finishReason, message } = choices[0]

      messages.push({ ...message, content: message.content! })

      const stepType = steps.length === 0
        ? 'initial'
        // eslint-disable-next-line sonarjs/no-nested-conditional
        : steps.at(-1)?.finishReason === 'tool-calls'
          ? 'tool-result'
          : 'continue'

      if ((message.content !== undefined && message.content.length > 0) || !message.tool_calls || steps.length >= (options.maxSteps ?? 1)) {
        const step: GenerateTextStepResult = {
          finishReason,
          stepType,
          text: message.content,
          toolCalls,
          toolResults,
          usage,
        }

        steps.push(step)

        if (options.onStepFinish)
          await options.onStepFinish(step)

        return {
          finishReason,
          messages,
          steps,
          text: message.content,
          toolCalls,
          toolResults,
          usage,
        }
      }

      for (const {
        function: { arguments: toolArgs, name: toolName },
        id: toolCallId,
        type: toolCallType,
      } of message.tool_calls) {
        const tool = options.tools?.find(tool => tool.function.name === toolName)

        if (!tool) {
          const availableTools = options.tools?.map(tool => tool.function.name)
          const availableToolsErrorMsg = availableTools === undefined
            ? 'No tools are available.'
            : `Available tools: ${availableTools.join(', ')}.`
          throw new Error(`Model tried to call unavailable tool '${toolName}. ${availableToolsErrorMsg}.`)
        }

        const parsedArgs = JSON.parse(toolArgs) as Record<string, unknown>
        const result = await tool.execute(parsedArgs, { abortSignal: options.abortSignal, messages, toolCallId })

        toolCalls.push({
          args: toolArgs,
          toolCallId,
          toolCallType,
          toolName,
        })

        toolResults.push({
          args: parsedArgs,
          result,
          toolCallId,
          toolName,
        })

        messages.push({
          content: result,
          role: 'tool',
          tool_call_id: toolCallId,
        })
      }

      const step: GenerateTextStepResult = {
        finishReason,
        stepType,
        text: message.content,
        toolCalls,
        toolResults,
        usage,
      }

      steps.push(step)

      if (options.onStepFinish)
        await options.onStepFinish(step)

      return async () => rawGenerateText({
        ...options,
        messages,
        steps,
      })
    })

export const generateText = async (options: GenerateTextOptions): Promise<GenerateTextResult> => {
  let result = await rawGenerateText(options)

  while (typeof result === 'function')
    result = await result()

  return result
}
