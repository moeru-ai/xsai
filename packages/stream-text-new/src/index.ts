import type { ChatOptions, Message, Tool, ToolCall, Usage } from '@xsai/shared-chat'

import { objCamelToSnake, trampoline } from '@xsai/shared'
import { chat, executeTool } from '@xsai/shared-chat'

import type { StreamTextEvent } from './types/event'

import { transformChunk } from './_transform-chunk'
import { DelayedPromise } from './internal/delayed-promise'

export interface StreamTextOptions extends ChatOptions {
  /** @default 1 */
  maxSteps?: number
  // TODO: onStepFinish, onFinish
  onEvent?: (event: StreamTextEvent) => Promise<unknown> | unknown
  /**
   * If you want to disable stream, use `@xsai/generate-{text,object}`.
   */
  stream?: never
  streamOptions?: {
    /**
     * Return usage.
     * @default `undefined`
     */
    includeUsage?: boolean
  }
  tools?: Tool[]
}

export interface StreamTextResult {
  fullStream: ReadableStream<StreamTextEvent>
  messages: Promise<Message[]>
  textStream: ReadableStream<string>
  usage: Promise<undefined | Usage>
  // TODO: steps, totalUsage
}

export const streamText = async (options: StreamTextOptions): Promise<StreamTextResult> => {
  // state
  const steps = []
  const messages: Message[] = structuredClone(options.messages)
  let usage: undefined | Usage

  // result state
  const resultMessages = new DelayedPromise<Message[]>()
  const resultUsage = new DelayedPromise<undefined | Usage>()

  // output
  let eventCtrl: ReadableStreamDefaultController<StreamTextEvent> | undefined
  let textCtrl: ReadableStreamDefaultController<string> | undefined
  const eventStream = new ReadableStream<StreamTextEvent>({ start: controller => eventCtrl = controller })
  const textStream = new ReadableStream<string>({ start: controller => textCtrl = controller })

  const pushEvent = (stepEvent: StreamTextEvent) => {
    eventCtrl?.enqueue(stepEvent)
    // eslint-disable-next-line sonarjs/void-use
    void options.onEvent?.(stepEvent)
  }

  const startStream = async () => {
    // let stepUsage: undefined | Usage
    const pushUsage = (u: Usage) => {
      usage = u
      // stepUsage = u
    }

    let text: string = ''
    const pushText = (content?: string) => {
      textCtrl?.enqueue(content)
      text += content
    }

    const toolCalls: ToolCall[] = []

    await chat({
      ...options,
      maxSteps: undefined,
      messages,
      stream: true,
      streamOptions: options.streamOptions != null
        ? objCamelToSnake(options.streamOptions)
        : undefined,
    }).then(async res => res.body!
      .pipeThrough(transformChunk())
      .pipeTo(new WritableStream({
        abort: (reason) => {
          eventCtrl?.error(reason)
          textCtrl?.error(reason)
        },
        close: () => {},
        write: (chunk) => {
          if (chunk.usage)
            pushUsage(chunk.usage)

          // skip if no choices
          if (chunk.choices == null || chunk.choices.length === 0)
            return

          const choice = chunk.choices[0]

          if (choice.delta.tool_calls?.length === 0 || choice.delta.tool_calls == null) {
            if (choice.delta.content != null) {
              pushEvent({ text: choice.delta.content, type: 'text-delta' })
              pushText(choice.delta.content)
            }
            else if (choice.delta.refusal != null) {
              pushEvent({ error: choice.delta.refusal, type: 'error' })
            }
            else if (choice.finish_reason != null && choice.finish_reason !== 'tool-calls') {
              pushEvent({ finishReason: choice.finish_reason, type: 'finish', usage })
            }
          }
          else {
            // https://platform.openai.com/docs/guides/function-calling?api-mode=chat&lang=javascript#streaming
            for (const toolCall of choice.delta.tool_calls) {
              const { index } = toolCall

              if (!toolCalls.at(index)) {
                toolCalls[index] = toolCall
                pushEvent({ toolCallId: toolCall.id, toolName: toolCall.function.name, type: 'tool-call-streaming-start' })
              }
              else {
                toolCalls[index].function.arguments += toolCall.function.arguments
                pushEvent({ argsTextDelta: toolCall.function.arguments, toolCallId: toolCall.id, toolName: toolCall.function.name, type: 'tool-call-delta' })
              }
            }
          }
        },
      })),
    )

    if (toolCalls.length === 0) {
      messages.push({ content: text, role: 'assistant' })
      return
    }

    for (const toolCall of toolCalls) {
      pushEvent({ args: toolCall.function.arguments, toolCallId: toolCall.id, toolName: toolCall.function.name, type: 'tool-call' })

      const { parsedArgs, result } = await executeTool({
        abortSignal: options.abortSignal,
        messages,
        toolCall,
        tools: options.tools,
      })

      pushEvent({ args: parsedArgs, result, toolCallId: toolCall.id, toolName: toolCall.function.name, type: 'tool-result' })

      // TODO: executeTool return message
      messages.push({
        content: result,
        role: 'tool',
        tool_call_id: toolCall.id,
      })
    }

    steps.push('TODO')

    if (steps.length < (options.maxSteps ?? 1))
      return async () => startStream()
  }

  try {
    await trampoline(async () => startStream())
  }
  finally {
    eventCtrl?.close()
    textCtrl?.close()

    resultMessages.resolve(messages)
    resultUsage.resolve(usage)
  }

  return {
    fullStream: eventStream,
    messages: resultMessages.promise,
    textStream,
    usage: resultUsage.promise,
  }
}
