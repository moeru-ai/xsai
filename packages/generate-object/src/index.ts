import { type Infer, type Schema, toJSONSchema, validate } from '@typeschema/main'
import { generateText, type GenerateTextOptions, type GenerateTextResult } from '@xsai/generate-text'
import { clean } from '@xsai/shared'

export interface GenerateObjectOptions<T extends Schema> extends GenerateTextOptions {
  schema: T
  schemaDescription?: string
  schemaName?: string
}

// TODO: toolCalls, toolResults
export interface GenerateObjectResult<T extends Schema> extends Omit<GenerateTextResult, 'text' | 'toolCalls' | 'toolResults'> {
  object: Infer<T>
}

/** @experimental WIP */
export const generateObject = async <T extends Schema>(options: GenerateObjectOptions<T>): Promise<GenerateObjectResult<T>> =>
  await generateText({
    ...options,
    response_format: {
      json_schema: {
        description: options.schemaDescription,
        name: options.schemaName ?? 'json_schema',
        schema: await toJSONSchema(options.schema)
          .then(json => clean({
            ...json,
            $schema: undefined,
          })),
        strict: true,
      },
      type: 'json_schema',
    },
    schema: undefined,
    schemaDescription: undefined,
    schemaName: undefined,
  })
    .then(async ({ finishReason, steps, text, usage }) => {
      const result = await validate(options.schema, JSON.parse(text!))

      if (result.success) {
        return {
          finishReason,
          object: result.data,
          steps,
          usage,
        }
      }
      else {
        throw new Error([
          'Schema validation failed:',
          ...result.issues.map(issue => `- ${issue.message}`),
        ].join('\n'))
      }
    })

export default generateObject
