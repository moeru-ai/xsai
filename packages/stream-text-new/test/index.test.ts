import { clean } from '@xsai/shared'
import { tool } from '@xsai/tool'
import { description, number, object, pipe } from 'valibot'
import { describe, expect, it } from 'vitest'

import type { StreamTextChunkResult } from '../src/_transform-chunk'

import { streamText } from '../src'

// make TS happy
// https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#browser_compatibility
declare global {
  interface ReadableStream<R = any> {
    // eslint-disable-next-line ts/method-signature-style
    [Symbol.asyncIterator](): AsyncIterableIterator<R>
  }
}

describe('@xsai/stream-text-new', async () => {
  const add = await tool({
    description: 'Adds two numbers',
    execute: ({ a, b }) => (a + b).toString(),
    name: 'weather',
    parameters: object({
      a: pipe(
        number(),
        description('First number'),
      ),
      b: pipe(
        number(),
        description('Second number'),
      ),
    }),
  })

  const cleanChunk = (chunk: StreamTextChunkResult) => clean({
    ...chunk,
    created: undefined,
    id: undefined,
  })

  it('chunkStream', async () => {
    const { chunkStream, textStream } = await streamText({
      baseURL: 'http://localhost:11434/v1/',
      maxSteps: 2,
      messages: [
        {
          content: 'You are a helpful assistant.',
          role: 'system',
        },
        {
          content: 'How many times does 114514 plus 1919810 equal? Please try to call the tool to solve the problem.',
          role: 'user',
        },
      ],
      model: 'PetrosStav/gemma3-tools:4b',
      seed: 114514,
      toolChoice: 'required',
      tools: [add],
    })

    const chunkResult = []
    for await (const chunk of chunkStream) {
      chunkResult.push(cleanChunk(chunk))
    }

    let textResult = ''
    for await (const text of textStream) {
      textResult += text
    }

    expect(chunkResult).toMatchSnapshot()
    expect(textResult).toMatchSnapshot()
  }, 30000)
})
