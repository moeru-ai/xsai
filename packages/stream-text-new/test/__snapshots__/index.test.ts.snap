// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@xsai/stream-text-new > fullStream 1`] = `
[
  {
    "toolName": "add",
    "type": "tool-call-streaming-start",
  },
  {
    "text": "",
    "type": "text-delta",
  },
  {
    "args": "{"a":114514,"b":1919810}",
    "toolName": "add",
    "type": "tool-call",
  },
  {
    "args": {
      "a": 114514,
      "b": 1919810,
    },
    "result": "2034324",
    "toolName": "add",
    "type": "tool-result",
  },
  {
    "text": "The",
    "type": "text-delta",
  },
  {
    "text": " sum",
    "type": "text-delta",
  },
  {
    "text": " of",
    "type": "text-delta",
  },
  {
    "text": " ",
    "type": "text-delta",
  },
  {
    "text": "1",
    "type": "text-delta",
  },
  {
    "text": "1",
    "type": "text-delta",
  },
  {
    "text": "4",
    "type": "text-delta",
  },
  {
    "text": "5",
    "type": "text-delta",
  },
  {
    "text": "1",
    "type": "text-delta",
  },
  {
    "text": "4",
    "type": "text-delta",
  },
  {
    "text": " and",
    "type": "text-delta",
  },
  {
    "text": " ",
    "type": "text-delta",
  },
  {
    "text": "1",
    "type": "text-delta",
  },
  {
    "text": "9",
    "type": "text-delta",
  },
  {
    "text": "1",
    "type": "text-delta",
  },
  {
    "text": "9",
    "type": "text-delta",
  },
  {
    "text": "8",
    "type": "text-delta",
  },
  {
    "text": "1",
    "type": "text-delta",
  },
  {
    "text": "0",
    "type": "text-delta",
  },
  {
    "text": " is",
    "type": "text-delta",
  },
  {
    "text": " ",
    "type": "text-delta",
  },
  {
    "text": "2",
    "type": "text-delta",
  },
  {
    "text": "0",
    "type": "text-delta",
  },
  {
    "text": "3",
    "type": "text-delta",
  },
  {
    "text": "4",
    "type": "text-delta",
  },
  {
    "text": "3",
    "type": "text-delta",
  },
  {
    "text": "2",
    "type": "text-delta",
  },
  {
    "text": "4",
    "type": "text-delta",
  },
  {
    "text": ".",
    "type": "text-delta",
  },
  {
    "text": "",
    "type": "text-delta",
  },
]
`;

exports[`@xsai/stream-text-new > fullStream 2`] = `"The sum of 114514 and 1919810 is 2034324."`;
